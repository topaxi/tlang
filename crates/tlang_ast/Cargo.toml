[package]
name = "tlang_ast"
version = "0.1.0"
edition.workspace = true

[dev-dependencies]
indoc = { workspace = true }
pretty_assertions = { workspace = true }
