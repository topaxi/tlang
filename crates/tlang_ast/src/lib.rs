pub mod node;
pub mod span;
pub mod symbols;
pub mod token;
