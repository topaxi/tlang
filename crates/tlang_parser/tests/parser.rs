use pretty_assertions::assert_eq;

use tlang_ast::{node, span::LineColumn, token::Literal};

mod common;

#[test]
fn test_unary_minus() {
    assert_parses!("-1;");
}

#[test]
fn test_simple_arithmetic_calculations() {
    assert_parser_snapshot!("1 + 2 + 3;");
    assert_parser_snapshot!("1 * 2 + 3;");
}

#[test]
fn test_simple_arithmetic_sum_mult_precedence() {
    assert_parser_snapshot!("1 + 2 * 3;");
}

#[test]
fn test_simple_arithmetic_sum_mult_precedence_parentheses() {
    assert_parser_snapshot!("(1 + 2) * 3;");
}

#[test]
fn test_simple_arithmetic_with_identifiers() {
    assert_parser_snapshot!("let x = 1; let y = 2; x + y;");
}

#[test]
fn test_simple_call() {
    assert_parser_snapshot!("foo(1, 2);");
}

#[test]
fn test_call_trailing_comma() {
    assert_parses!("foo(1, 2,);");
}

#[test]
#[should_panic]
fn test_call_list_mandatory_comma() {
    assert_parses!("foo(1 2);");
}

#[test]
fn test_nested_call() {
    assert_parser_snapshot!("foo(bar(1), 2);");
}

#[test]
fn test_call_with_expression() {
    assert_parser_snapshot!("foo(1 + 2, 3);");
}

#[test]
fn test_block_expression() {
    assert_parser_snapshot!("let x = { 1 + 2; };");
}

#[test]
fn test_if_statement() {
    assert_parser_snapshot!("if (1 + 2) { 3 + 4; }");
}

#[test]
fn test_if_else_statement() {
    assert_parser_snapshot!("if true { 1; } else { 2; }");
    assert_parser_snapshot!("if (1 + 2) { 3 + 4; } else { 5 + 6; }");
}

#[test]
fn test_if_else_if_statement() {
    assert_parser_snapshot!("if true { 1; } else if false { 2; }");
}

#[test]
fn test_if_else_as_last_expression() {
    assert_parser_snapshot!("fn main() { if true { 1 } else { 2 } }");
}

#[test]
fn test_if_expression() {
    assert_parser_snapshot!("let x = if (true) { 1; } else { 2; };");
}

#[test]
fn test_stray_semicolon() {
    assert_parses!("1 + 2;;");
}

#[test]
fn test_exponentiation() {
    assert_parser_snapshot!("1 * 2 ** 3;");
    assert_parser_snapshot!("1 ** 2 * 3;");
    assert_parser_snapshot!("1 ** 2 ** 3;");
}

#[test]
fn test_pattern_matching() {
    assert_parser_snapshot!("let x = match 1 { 1 => 2, 3 => 4, _ => 5 };");
}

#[test]
fn test_list_literal() {
    assert_parser_snapshot!("let x = [];");
    assert_parser_snapshot!("let x = [1, 2, 3];");
}

#[test]
fn test_list_literal_with_trailing_comma() {
    assert_parser_snapshot!("let x = [1, 2, 3,];");
}

#[test]
fn test_single_line_comments() {
    assert_parser_snapshot!("// this is a comment");
}

#[test]
fn test_multi_line_comments() {
    assert_parser_snapshot!("/* this is a comment */");
    assert_parser_snapshot!("/* this is a comment\n   spanning multiple lines */");
}

#[test]
fn test_string_literal() {
    let program = parse!(r#""foo";"#);

    assert_eq!(
        program,
        node::new!(Module(vec![node::new!(
            ExpressionStatement(Box::new(node::new!(
                Literal(Literal::String("foo".to_string())),
                Span::new(
                    LineColumn { line: 0, column: 0 },
                    LineColumn { line: 0, column: 5 }
                )
            ))),
            Span::new(
                LineColumn { line: 0, column: 0 },
                LineColumn { line: 0, column: 6 }
            )
        )]))
    );
}

#[test]
fn test_char_literal() {
    let program = parse!(r#"'a';"#);

    assert_eq!(
        program,
        node::new!(Module(vec![node::new!(
            ExpressionStatement(Box::new(node::new!(
                Literal(Literal::Char("a".to_string())),
                Span::new(
                    LineColumn { line: 0, column: 0 },
                    LineColumn { line: 0, column: 3 }
                )
            ))),
            Span::new(
                LineColumn { line: 0, column: 0 },
                LineColumn { line: 0, column: 4 }
            )
        )]))
    );
}

#[test]
fn test_dictionary_literal() {
    assert_parser_snapshot!(r#"let x = { foo: 1, bar: 2 };"#);
}

#[test]
fn test_function_call_with_dictionary_no_parens() {
    assert_parser_snapshot!(r#"foo { foo: 1, bar: 2 };"#);
}

#[test]
fn test_call_return_value() {
    assert_parser_snapshot!("foo()();");
}

#[test]
fn test_field_access_expressions() {
    assert_parser_snapshot!("foo.bar;");
    assert_parser_snapshot!("foo.bar.baz;");
}

#[test]
fn test_dynamic_index_access() {
    assert_parser_snapshot!("let x = foo[1];");
    assert_parser_snapshot!("let x = foo[1][2];");
}
