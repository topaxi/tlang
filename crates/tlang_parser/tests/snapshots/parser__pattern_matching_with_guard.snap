---
source: crates/tlang_parser/tests/parser.rs
expression: "parse! (\"let x = match 1 { a if a == 1 => 2, _ => 3 };\")"
---
Module(
  id: NodeId(1),
  statements: [
    Stmt(
      id: NodeId(14),
      kind: Let(LetDeclaration(
        pattern: Pat(
          id: NodeId(2),
          kind: Identifier(Ident(
            name: "x",
            span: Span(
              start: LineColumn(
                line: 0,
                column: 4,
              ),
              end: LineColumn(
                line: 0,
                column: 5,
              ),
            ),
          )),
          leading_comments: [],
          trailing_comments: [],
          span: Span(
            start: LineColumn(
              line: 0,
              column: 4,
            ),
            end: LineColumn(
              line: 0,
              column: 5,
            ),
          ),
        ),
        expression: Expr(
          id: NodeId(13),
          kind: Match(MatchExpression(
            expression: Expr(
              id: NodeId(3),
              kind: Literal(UnsignedInteger(1)),
              leading_comments: [],
              trailing_comments: [],
              span: Span(
                start: LineColumn(
                  line: 0,
                  column: 14,
                ),
                end: LineColumn(
                  line: 0,
                  column: 15,
                ),
              ),
            ),
            arms: [
              MatchArm(
                id: NodeId(9),
                pattern: Pat(
                  id: NodeId(4),
                  kind: Identifier(Ident(
                    name: "a",
                    span: Span(
                      start: LineColumn(
                        line: 0,
                        column: 18,
                      ),
                      end: LineColumn(
                        line: 0,
                        column: 19,
                      ),
                    ),
                  )),
                  leading_comments: [],
                  trailing_comments: [],
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 18,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 19,
                    ),
                  ),
                ),
                guard: Some(Expr(
                  id: NodeId(7),
                  kind: BinaryOp(BinaryOpExpression(
                    op: Equal,
                    lhs: Expr(
                      id: NodeId(5),
                      kind: Path(Path(
                        segments: [
                          Ident(
                            name: "a",
                            span: Span(
                              start: LineColumn(
                                line: 0,
                                column: 23,
                              ),
                              end: LineColumn(
                                line: 0,
                                column: 24,
                              ),
                            ),
                          ),
                        ],
                        span: Span(
                          start: LineColumn(
                            line: 0,
                            column: 23,
                          ),
                          end: LineColumn(
                            line: 0,
                            column: 24,
                          ),
                        ),
                      )),
                      leading_comments: [],
                      trailing_comments: [],
                      span: Span(
                        start: LineColumn(
                          line: 0,
                          column: 23,
                        ),
                        end: LineColumn(
                          line: 0,
                          column: 24,
                        ),
                      ),
                    ),
                    rhs: Expr(
                      id: NodeId(6),
                      kind: Literal(UnsignedInteger(1)),
                      leading_comments: [],
                      trailing_comments: [],
                      span: Span(
                        start: LineColumn(
                          line: 0,
                          column: 28,
                        ),
                        end: LineColumn(
                          line: 0,
                          column: 29,
                        ),
                      ),
                    ),
                  )),
                  leading_comments: [],
                  trailing_comments: [],
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 23,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 29,
                    ),
                  ),
                )),
                expression: Expr(
                  id: NodeId(8),
                  kind: Literal(UnsignedInteger(2)),
                  leading_comments: [],
                  trailing_comments: [],
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 33,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 34,
                    ),
                  ),
                ),
              ),
              MatchArm(
                id: NodeId(12),
                pattern: Pat(
                  id: NodeId(10),
                  kind: Wildcard,
                  leading_comments: [],
                  trailing_comments: [],
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 36,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 37,
                    ),
                  ),
                ),
                guard: None,
                expression: Expr(
                  id: NodeId(11),
                  kind: Literal(UnsignedInteger(3)),
                  leading_comments: [],
                  trailing_comments: [],
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 41,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 42,
                    ),
                  ),
                ),
              ),
            ],
          )),
          leading_comments: [],
          trailing_comments: [],
          span: Span(
            start: LineColumn(
              line: 0,
              column: 8,
            ),
            end: LineColumn(
              line: 0,
              column: 44,
            ),
          ),
        ),
        type_annotation: None,
      )),
      span: Span(
        start: LineColumn(
          line: 0,
          column: 0,
        ),
        end: LineColumn(
          line: 0,
          column: 45,
        ),
      ),
      leading_comments: [],
      trailing_comments: [],
    ),
  ],
  span: Span(
    start: LineColumn(
      line: 0,
      column: 0,
    ),
    end: LineColumn(
      line: 0,
      column: 0,
    ),
  ),
)
